src/commissaire_service/clusterexec/__init__.py

```diff
@@ -0,0 +1,305 @@
+# Copyright (C) 2016  Red Hat, Inc
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import datetime
+
+from commissaire.models import (
+    ClusterDeploy, ClusterUpgrade, ClusterRestart, Cluster, Hosts)
+from commissaire.util.ssh import TemporarySSHKey
+
+from commissaire_service.oscmd import get_oscmd
+from commissaire_service.service import CommissaireService
+from commissaire_service.transport import ansibleapi
+
+
+class ClusterExecService(CommissaireService):
+    """
+    Executes operations over a cluster by way of remote shell commands.
+    """
+
+    def __init__(self, exchange_name, connection_url):
+        """
+        Creates a new ClusterExecService.
+
+        :param exchange_name: Name of the topic exchange
+        :type exchange_name: str
+        :param connection_url: Kombu connection URL
+        :type connection_url: str
+        """
+        queue_kwargs = [
+            {'routing_key': 'jobs.clusterexec'}
+        ]
+        super().__init__(exchange_name, connection_url, queue_kwargs)
+
+    def _execute(self, model_instance, command_args, finished_hosts_key):
+        """
+        Remotely executes OS-specific shell commands across a cluster.
+
+        :param model_instance: Initial model for the async operation
+        :type model_instance: commissaire.models.Model
+        :param command_args: Command name + arguments as a tuple
+        :type command_args: tuple
+        :param finished_hosts_key: Model attribute name for finished hosts
+        :type finished_hosts_key: str
+        """
+        # Split out the command name.
+        command_name = command_args[0]
+        command_args = command_args[1:]
+
+        end_status = 'finished'
+
+        # XXX We assume the model instance names a cluster.
+        #     Note, cluster_name is used in the except clause,
+        #     so it must be reliably defined.
+        cluster_name = getattr(model_instance, 'name', None)
+
+        try:
+            # Set the initial status in the store.
+            assert cluster_name is not None
+            model_json_data = model_instance.to_json()
+            self.logger.info('Setting initial status.')
+            self.logger.debug('Status={}'.format(model_json_data))
+            params = {
```

Comment:
nit: This line to 83 are reused more than a few times. It might be worth turning it into a function in this service. I won't block on it though.